edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs

```diff
@@ -263,33 +264,55 @@ void InternalConnectionStatusChangesHandler(ConnectionStatus status, ConnectionS
         async Task<string> GetNewToken(string iotHub, string id, string currentToken, IIdentity currentIdentity)
         {
             Events.GetNewToken(id);
-            // We have to catch UnauthorizedAccessException, because on IsTokenUsable, we call parse from
-            // Device Client and it throws if the token is expired.
-            if (IsTokenUsable(iotHub, currentToken))
-            {
-                Events.UsingExistingToken(id);
-                return currentToken;
-            }
-            else
+            bool retrying = false;
+            string token = currentToken;
+            while (true)
             {
-                Events.TokenExpired(id, currentToken);
-            }
+                // We have to catch UnauthorizedAccessException, because on IsTokenUsable, we call parse from
+                // Device Client and it throws if the token is expired.
+                if (IsTokenUsable(iotHub, token))
+                {
+                    if (retrying)
+                    {
+                        Events.NewTokenObtained(iotHub, id, token);
+                    }
+                    else
+                    {
+                        Events.UsingExistingToken(id);
+                    }
+                    return token;
+                }
+                else
+                {
+                    Events.TokenNotUsable(iotHub, id, token);                    
+                }
 
-            // No need to lock here as the lock is being held by the refresher.
-            TaskCompletionSource<string> tcs = this.tokenGetter
-                .GetOrElse(
-                    () =>
+                bool newTokenGetterCreated = false;
+                // No need to lock here as the lock is being held by the refresher.
+                TaskCompletionSource<string> tcs = this.tokenGetter
+                    .GetOrElse(
+                        () =>
+                        {
+                            Events.SafeCreateNewToken(id);
+                            var taskCompletionSource = new TaskCompletionSource<string>();
+                            this.tokenGetter = Option.Some(taskCompletionSource);
+                            newTokenGetterCreated = true;                            
+                            return taskCompletionSource;
+                        });
+
+                if (newTokenGetterCreated)
+                {
+                    if (retrying)
```

Comment:
It is not clear to me when and how many times 'newTokenGetterCreated' and 'retrying' would both be true. Perhaps some comments could help clarify when tokenGetters are set and cleared.